blockstack-vm/src/contexts.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use super::types::{DefinedFunction, FunctionIdentifier, Value};
+use super::database::ContractDatabase;
+
+pub struct Environment <'a> {
+    pub global_context: Context <'a>,
+    pub call_stack: CallStack,
+    pub database: Box<ContractDatabase>
+}
+
+impl <'a> Environment <'a> {
+    pub fn new(database: Box<ContractDatabase>) -> Environment<'a> {
+        let global_context = Context::new();
+        Environment {
+            global_context: global_context,
+            call_stack: CallStack::new(),
+            database: database
+        }
+    }
+}
+
+pub struct Context <'a> {
+    pub parent: Option< &'a Context<'a>>,
+    pub variables: HashMap<String, Value>,
+    pub functions: HashMap<String, Box<DefinedFunction>>,
+}
+
+impl <'a> Context <'a> {
+    pub fn new() -> Context<'a> {
+        Context { parent: Option::None,
+                  variables: HashMap::new(),
+                  functions: HashMap::new() }
+    }
+    
+    pub fn extend(&'a self) -> Context<'a> {
+        Context {
+            parent: Some(self),
+            variables: HashMap::new(),
+            functions: HashMap::new()
+        }
+    }
+
+    pub fn lookup_variable(&self, name: &str) -> Option<Value> {
+        match self.variables.get(name) {
+            Some(value) => Option::Some((*value).clone()),
+            None => {
+                match self.parent {
+                    Some(parent) => parent.lookup_variable(name),
+                    None => Option::None
+                }
+            }
+        }
+    }
+
+    pub fn lookup_function(&self, name: &str) -> Option<Box<DefinedFunction>> {
+        match self.functions.get(name) {
+            Some(value) => {
+                Option::Some(Box::new(*value.clone()))
```

Comment:
Why are we heap-allocating this?  Just curious -- it looks like we could get away with returning an immutable reference.

Reply:
Ah, yeah, it could be an immutable reference with the same lifetime as the Context object.

Reply:
Oh -- nevermind, I think trying to return an immutable reference here is hard (even with the lifetime genericized with the Context object). User defined functions here are stored in a Map object, which even though we never delete keys from, rust will not let the references outlive the function call.

Reply:
Gotcha.  Why use a `Box` at all then?  Why not just clone it and return it?

Reply:
But I don't actually need the Box at all, since DefinedFunctions have known compile-time size -- I think in an early version they didn't. I removed the Box wrapper in my branch.